contracts/QuestNFT.sol

```diff
@@ -0,0 +1,219 @@
+// SPDX-License-Identifier: UNLICENSED
+pragma solidity =0.8.16;
+
+import {Initializable} from '@openzeppelin/contracts-upgradeable/proxy/utils/Initializable.sol';
+import {ERC721Upgradeable} from "@openzeppelin/contracts-upgradeable/token/ERC721/ERC721Upgradeable.sol";
+import {PausableUpgradeable} from '@openzeppelin/contracts-upgradeable/security/PausableUpgradeable.sol';
+import {OwnableUpgradeable} from '@openzeppelin/contracts-upgradeable/access/OwnableUpgradeable.sol';
+import {CountersUpgradeable} from "@openzeppelin/contracts-upgradeable/utils/CountersUpgradeable.sol";
+import {IERC2981Upgradeable, IERC165Upgradeable} from '@openzeppelin/contracts-upgradeable/interfaces/IERC2981Upgradeable.sol';
+import {SafeERC20, IERC20} from '@openzeppelin/contracts/token/ERC20/utils/SafeERC20.sol';
+import {Base64} from '@openzeppelin/contracts/utils/Base64.sol';
+import {Strings} from '@openzeppelin/contracts/utils/Strings.sol';
+import {ReentrancyGuardUpgradeable} from "@openzeppelin/contracts-upgradeable/security/ReentrancyGuardUpgradeable.sol";
+
+/// @title QuestNFT
+/// @author RabbitHole.gg
+/// @notice This contract is the Erc721 Quest Completion contract. It is the NFT that can be minted after a quest is completed.
+contract QuestNFT is Initializable, ERC721Upgradeable, PausableUpgradeable, OwnableUpgradeable, IERC2981Upgradeable, ReentrancyGuardUpgradeable {
+    using CountersUpgradeable for CountersUpgradeable.Counter;
+    using SafeERC20 for IERC20;
+
+    CountersUpgradeable.Counter private _tokenIdCounter;
+
+    address public protocolFeeRecipient;
+    address public minterAddress;
+    uint256 public endTime;
+    uint256 public startTime;
+    uint256 public totalParticipants;
+    uint256 public questFee;
+    string public jsonSpecCID;
+    string public imageIPFSHash;
+    string public description;
+    string public questId;
+
+    event JsonSpecCIDSet(string cid);
+
+    /// @custom:oz-upgrades-unsafe-allow constructor
+    constructor() {
+        _disableInitializers();
+    }
+
+    function initialize(
+        uint256 endTime_,
+        uint256 startTime_,
+        uint256 totalParticipants_,
+        string memory questId_,
+        uint256 questFee_,
+        address protocolFeeRecipient_,
+        address minterAddress_, // should always be the QuestFactory contract
+        string memory jsonSpecCID_,
+        string memory name_,
+        string memory symbol_,
+        string memory description_,
+        string memory imageIPFSHash_
+    ) external initializer {
+        require (endTime_ > block.timestamp, 'endTime_ in past');
```

Comment:
Extremely extremely rare edge case, but if `block.timestamp` == `endTime_` the error message is a little inaccurate